### PR TITLE
[crowdstrike] Paginate indicator requests

### LINKED ISSUE #4497: Crowdstrike Connector not paginating results
## Environment

1. OS (where OpenCTI server runs): Mac OS 10
2. OpenCTI version: 6.7.11
3. OpenCTI client: frontend
4. Other environment details: Running locally with `docker compose` and have set-up the crowdstrike connector.

## Reproducible Steps

Steps to create the smallest reproducible scenario:
1. Run OpenCTI
2. Configure Crowdstrike Connector (will require their API Key)
3. Notice an entry like this in the logs:
> "Indicators fetched to be processed (Crowdstrike max limit = 10000)", "attributes": {"resources_count": 1000, "remaining_resources": 31093625}}

## Expected Output

There is a hard-coded limit in the codebase to fetch only 1000 items at a time, which is reasonable but will be practically impossible to catch-up with everything. The proposed fix is to somehow use pagination to fetch all the data **or** refactor the code to drop the `TIMESTAMP` variables and instead use a `now` type of timestamp when the connectors first starts. If you're looking to fetch data in the past (and doesn't have to be huge, like 6 months) it will never actually have all the indicators up-to-date.

## Actual Output

Only 1000 indicators are processed at a time.

## Additional information

N/A 

## Screenshots (optional)

N/A

### PR DESCRIPTION
**Pull request by gsilvapt**
### Proposed changes

* `IndicatorsAPI.get_combined_indicator_entities` now accepts a `next_page` continuation token and re-exposes the next-page token of the current response as `response_body["next_page"]` (a single string, not a parsed query-string dict, so the contract is unambiguous).
* `IndicatorImporter._fetch_indicators` is rewritten around a `_paginated_query_indicators` helper that walks every page returned by the API while honouring a configurable `max_records_per_run` cap. The cap is enforced at the resource level so a single page is never partially yielded with a dangling truncation, and pagination is robust to `meta.pagination` being missing or `None`.
* New `CROWDSTRIKE_INDICATOR_MAX_RECORDS_PER_RUN` setting (default `20000`) bounds the per-run volume. CrowdStrike can return up to ~1.5M IoCs over 30 days which would overwhelm the OpenCTI ingestion pipeline; set to `0` (or a negative number) to disable the cap entirely.
* The setting is exposed in `config.yml.sample`, `__metadata__/connector_config_schema.json` and `__metadata__/CONNECTOR_CONFIG_DOC.md`.
* New `tests/test_indicator_pagination.py` (11 unit tests) covers multi-page walking, stop-on-empty, the `meta.pagination == None` edge case, the per-run cap (within a single page and across multiple pages), the disabled cap (`None` / `0` / `-1`) and the `get_next_page` token parser.

### Related issues

* Closes #4497

### Checklist

- [x] I consider the submitted work as finished
- [x] I have signed my commits using GPG key.
- [x] I tested the code for its functionality using different use cases
- [x] I added/update the relevant documentation (either on github or on notion)
- [x] Where necessary I refactored code to improve the overall quality

### Further comments

PR rebased onto the current master (the connector layout evolved while this PR was open: `connector.py`, `RelatedActorImporter`, `AttackTechniqueLookup`, `indicator_max_age_by_type`, …) and squashed into a single GPG-signed commit. Every Copilot thread is addressed and resolved, and the per-run cap requested by @Powlinett is now wired through the pydantic settings, the env-var schema and the YAML config sample.

Local validation:
- `isort==6.0.1 --profile black` (clean)
- `black==25.1.0 --check` (clean)
- `flake8 --ignore=E,W` (clean)
- `pylint --disable=all --enable=no_generated_id_stix,no-value-for-parameter,unused-import` with the `linter_stix_id_generator` plugin → 10/10
- `pytest external-import/crowdstrike/tests/` → **141 passed**

### CODE REVIEW

**Comment by gsilvapt** [comment]
> Hi @gsilvapt ,
> 
> Thank you for your contribution 🙏
> 
> I tested the pagination and it seems to work as expected 👍 Could you resolve the conflict on the PR so I can review it one last time and merge it please? 🙏
> 
> Feel free to ping me for any help 😇

Thank you for reviewing! Apologies for the late response.
Just solved the merge conflicts for a final round of tests 🙌 

Let me know if you need any further help.

File: external-import/crowdstrike/src/crowdstrike_feeds_connector/indicator/importer.py [hunk 1]
```diff
@@ -90,82 +90,136 @@ def run(self, state: Dict[str, Any]) -> Dict[str, Any]:
...
+        next_page_params: Optional[Dict] = None,
+    ) -> Dict[str, Any]:
+        """Query a single page of indicators from the CrowdStrike API."""
+        next_page = None
+        if next_page_params and "next_page" in next_page_params:
+            next_page = (
+                next_page_params["next_page"][0]
+                if isinstance(next_page_params["next_page"], list)
+                else next_page_params["next_page"]
             )
```
[the remark this answers is not in the document]
Comment (SamuelHassine) [reply external-import/crowdstrike/src/crowdstrike_feeds_connector/indicator/importer.py]: Fixed: the client now returns `next_page` as a single string token (not a `parse_qs` dict), so there is no ambiguity left. The importer forwards that token as-is on the next call and stops when it is `None` / empty.

File: external-import/crowdstrike/src/crowdstrike_feeds_connector/indicator/importer.py [hunk 2]
```diff
@@ -90,82 +90,136 @@ def run(self, state: Dict[str, Any]) -> Dict[str, Any]:
...
+            # Yield each indicator
+            yield from resources
+
+            # Check for next page
+            next_page_details = response.get("next_page_details")
+            if not next_page_details:
+                break
+
+            next_page_params = next_page_details
+
```
[the remark this answers is not in the document]
Comment (SamuelHassine) [reply external-import/crowdstrike/src/crowdstrike_feeds_connector/indicator/importer.py]: Done: `tests/test_indicator_pagination.py` (11 unit tests) covers multi-page walking, stopping when the next-page token is missing, the per-run cap (within a single page and across multiple pages), the disabled-cap variants, the `meta.pagination is None` edge case, and the `get_next_page` token parser. All 141 tests in `external-import/crowdstrike/tests/` pass locally.

File: external-import/crowdstrike/src/crowdstrike_feeds_services/client/indicators.py [hunk 3]
```diff
@@ -9,19 +10,33 @@ def __init__(self, helper):
...
+            "deep_pagination": deep_pagination,
+        }
+
+        if next_page:
+            kwargs["next_page"] = next_page
+
+        response = self.cs_intel.query_indicator_entities(**kwargs)
 
         response_body = response["body"]
         response_body["next_page_details"] = None
```
[the remark this answers is not in the document]
Comment (SamuelHassine) [reply external-import/crowdstrike/src/crowdstrike_feeds_services/client/indicators.py]: Fixed: `get_combined_indicator_entities` now sets `response_body['next_page']` to a single string token (the value of the `next_page` query parameter extracted from the upstream `Next-Page` header) or `None`. `IndicatorsAPI.get_next_page` is typed `Optional[str]` accordingly. Callers just need to forward that string back as the `next_page=` kwarg, which is what `_paginated_query_indicators` does.

File: external-import/crowdstrike/src/crowdstrike_feeds_connector/indicator/importer.py [hunk 4]
```diff
@@ -90,82 +90,136 @@ def run(self, state: Dict[str, Any]) -> Dict[str, Any]:
...
+            if not response or not response.get("resources"):
+                break
+
+            resources = response["resources"]
+            batch_size = len(resources)
+            total_fetched += batch_size
+
+            # Log progress
+            meta = response.get("meta", {})
+            pagination = meta.get("pagination", {})
```
[the remark this answers is not in the document]
Comment (SamuelHassine) [reply external-import/crowdstrike/src/crowdstrike_feeds_connector/indicator/importer.py]: Fixed: `_paginated_query_indicators` now uses `meta.get('pagination') or {}` (and `meta = response.get('meta') or {}` for good measure), so the loop is safe whether `meta` / `pagination` is missing, `None` or an empty dict. There is also a dedicated unit test (`test_pagination_handles_missing_pagination_metadata`) covering the `meta.pagination == None` case.

**Comment by SamuelHassine** [comment]
@Powlinett The max-IoCs-per-run threshold is now in. `CROWDSTRIKE_INDICATOR_MAX_RECORDS_PER_RUN` defaults to `20000` (your suggested value) but can be tuned via env / `config.yml` and disabled with `0` for users who want every IoC. Pagination logic refactored accordingly, all four Copilot threads addressed, and the change is covered by 11 new unit tests in `tests/test_indicator_pagination.py` (full suite still passes: 141 tests).

**Review by SamuelHassine — DISMISSED** [review]
Full review done.

Rebased the change set onto the current master (the connector layout has evolved significantly since this PR was opened: new `connector.py`, `RelatedActorImporter`, `AttackTechniqueLookup`, `indicator_max_age_by_type`, …) and squashed everything into a single GPG-signed commit. All four Copilot review threads have been addressed and resolved, the per-run cap requested by @Powlinett is now wired through, and the change is now covered by 11 new unit tests.

Highlights:
- `IndicatorsAPI.get_combined_indicator_entities` now accepts a `next_page` continuation token and re-exposes the next-page token of the current response as `response_body["next_page"]` (a single string, not a parsed query-string dict). `get_next_page` is typed `Optional[str]` accordingly.
- `IndicatorImporter._fetch_indicators` is rewritten around `_paginated_query_indicators`, which walks every page until either the upstream API returns no more or the configured cap is reached. The cap is enforced at the resource level (never partial-page truncation), and the loop is robust to `meta.pagination` being missing or `None`.
- New `CROWDSTRIKE_INDICATOR_MAX_RECORDS_PER_RUN` setting (default `20000`) bounds the per-run volume; set to `0` (or a negative number) to disable the cap entirely. Exposed in `config.yml.sample`, the JSON schema and the auto-generated docs.
- New `tests/test_indicator_pagination.py` (11 unit tests) covers the multi-page walk, stop-on-empty, the `meta.pagination == None` edge case, the per-run cap (within a single page and across multiple pages), the three disabled-cap variants (`None` / `0` / `-1`) and the `get_next_page` token parser.

CI is fully green on the GitHub Actions side (`Test external-import/crowdstrike`, `Test tests/test-requirements.txt`, all `PR conventions checks`, `Baseline coverage`, `Do Not Merge label check`, `codecov/patch`, `codecov/project`). The CircleCI workflows haven't fired yet on this revision but all 141 tests in `external-import/crowdstrike/tests/` pass locally with `isort==6.0.1` / `black==25.1.0` / `flake8` / `pylint --enable=no_generated_id_stix,no-value-for-parameter,unused-import` clean (pylint 10/10).

The only remaining unstable status is `filigran/cla` (waiting for @gsilvapt to sign), which can be overridden at merge time.

**Comment by gsilvapt** [comment]
Just signed it to unblock you @SamuelHassine , even though you can overwrite it apparently. Thanks for bumping this!

File: external-import/crowdstrike/src/crowdstrike_feeds_connector/indicator/importer.py [hunk 5]
```diff
@@ -91,6 +92,11 @@ def __init__(self, config: IndicatorImporterConfig) -> None:
         self.scopes = config.scopes
```
[the remark this answers is not in the document]
Comment (SamuelHassine) [reply external-import/crowdstrike/src/crowdstrike_feeds_connector/indicator/importer.py:91]: Fixed: removed the unused `self.next_page` attribute from `IndicatorImporter.__init__`. It was leftover from the previous pagination implementation - the actual cursor lives in the local `next_page_token` inside `_paginated_query_indicators`, and nothing else read this attribute. `pylint --enable=unused-import` and the existing 141-test suite still pass.

File: external-import/crowdstrike/src/crowdstrike_feeds_connector/indicator/importer.py [hunk 6]
```diff
@@ -138,54 +144,117 @@ def run(self, state: Dict[str, Any]) -> Dict[str, Any]:
...
+        self, limit: int, sort: str, fql_filter: str
+    ) -> List[dict]:
+        """Walk every page of the indicator API and return the aggregated list.
+
+        The continuation token is the ``next_page`` string returned by the
+        upstream API and forwarded as-is by :class:`IndicatorsAPI`. When
+        ``max_records_per_run`` is set, pagination stops as soon as that many
+        indicators have been collected; the cap is enforced at the resource
+        level so a single page is never partially yielded with a dangling
+        truncation.
```
[the remark this answers is not in the document]
Comment (SamuelHassine) [reply external-import/crowdstrike/src/crowdstrike_feeds_connector/indicator/importer.py]: Fixed: the docstring of `_paginated_query_indicators` has been rewritten to match the implementation. It now states that the cap is enforced at the resource level by slicing the *last* page to exactly the remaining quota, so the aggregated batch contains at most `max_records_per_run` records and we never yield a record beyond the cap. Slicing only happens on the page that crosses the cap boundary; previous pages are always returned in full, which is what `test_pagination_truncates_within_a_single_page` / `test_pagination_truncates_across_multiple_pages` already verify.

File: external-import/crowdstrike/src/config.yml.sample [hunk 7]
```diff
@@ -45,5 +45,6 @@ crowdstrike:
   indicator_high_score: 80
   indicator_high_score_labels: 'MaliciousConfidence/High'
   indicator_unwanted_labels: '' # Can be used to filter low confidence indicators: "MaliciousConfidence/Low", "MaliciousConfidence/Medium"
+  indicator_max_records_per_run: 20000 # Max indicators ingested per run (CrowdStrike can return up to ~1.5M IoCs over 30 days). Set to 0 to disable the cap.
```
[the remark this answers is not in the document]
Comment (SamuelHassine) [reply external-import/crowdstrike/src/config.yml.sample]: Fixed: the inline comment in `src/config.yml.sample` now reads `Set to 0 (or any negative number) to disable the cap.`, matching what `__metadata__/CONNECTOR_CONFIG_DOC.md` and `__metadata__/connector_config_schema.json` already say. So all three docs are now consistent with the runtime behavior (`config.max_records_per_run if config.max_records_per_run and config.max_records_per_run > 0 else None` in the importer).

**Review by SamuelHassine — DISMISSED** [review]
Re-approving after pushing review-comment cleanup (removed unused self.next_page, aligned docstring with slicing behavior, mentioned negative-cap-disables in config.yml.sample). All 7 review threads are resolved; locally validated with isort/black/flake8, pylint+linter_stix_id_generator (10/10), and pytest external-import/crowdstrike/tests/ (141 passed).

File: external-import/crowdstrike/tests/test_indicator_pagination.py [hunk 8]
```diff
@@ -0,0 +1,218 @@
...
+os.environ.setdefault("OPENCTI_TOKEN", "token")
+os.environ.setdefault("CONNECTOR_ID", "ChangeMe")
+os.environ.setdefault("CONNECTOR_TYPE", "EXTERNAL_IMPORT")
+os.environ.setdefault("CONNECTOR_NAME", "CrowdStrike")
+os.environ.setdefault("CONNECTOR_SCOPE", "crowdstrike")
+os.environ.setdefault("CONNECTOR_LOG_LEVEL", "info")
+os.environ.setdefault("CONNECTOR_DURATION_PERIOD", "PT30M")
+os.environ.setdefault("CROWDSTRIKE_CLIENT_ID", "ChangeMe")
+os.environ.setdefault("CROWDSTRIKE_CLIENT_SECRET", "ChangeMe")
+
```
[the remark this answers is not in the document]
Comment (SamuelHassine) [reply external-import/crowdstrike/tests/test_indicator_pagination.py]: Fixed: the module-level `os.environ.setdefault(...)` block is gone. The env vars are now driven by a function-scoped `autouse` fixture (`_crowdstrike_env`) that uses `monkeypatch.setenv` ? which always **overrides** any value the test runner might have inherited (whereas `setdefault` would have silently kept an invalid one) ? and pytest restores the original env at teardown so the settings can no longer leak into the other CrowdStrike test modules.

Concretely:

```python
_REQUIRED_ENV: dict[str, str] = {
    "OPENCTI_URL": "http://localhost:8080",
    ...
}

@pytest.fixture(autouse=True)
def _crowdstrike_env(monkeypatch: pytest.MonkeyPatch) -> None:
    for key, value in _REQUIRED_ENV.items():
        monkeypatch.setenv(key, value)
```

The `crowdstrike_feeds_connector` / `crowdstrike_feeds_services` imports are also now **lazy** (inside `_build_importer` and the three `test_get_next_page_*` cases) so the env fixture has applied before the connector's pydantic settings are constructed at module import time.

All 141 tests in `external-import/crowdstrike/tests/` continue to pass locally.

**Review by SamuelHassine — APPROVED** [review]
Re-approving after replacing the module-level os.environ.setdefault block with an autouse monkeypatch.setenv fixture (overrides invalid runner values + cleans up at teardown) and lazy-importing the connector modules inside _build_importer so the fixture has applied before pydantic-settings runs. All 141 tests in external-import/crowdstrike/tests/ still pass; isort/black/flake8 clean.
